File: day1/objc/day1-part1.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *filePath = @"../input.txt";
        NSString *fileContents = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
        
        NSArray *pairs = [fileContents componentsSeparatedByString:@"\n"];
        
        NSMutableArray *list1 = [NSMutableArray array];
        NSMutableArray *list2 = [NSMutableArray array];
        
        for (NSString *pair in pairs) {
            NSArray *components = [pair componentsSeparatedByString:@"   "];
            
            [list1 addObject:@([components[0] integerValue])];
            [list2 addObject:@([components[1] integerValue])];
        }
        
        NSComparisonResult (^ascendingComparison)(NSNumber *num1, NSNumber *num2) = ^NSComparisonResult(NSNumber *num1, NSNumber *num2) {
            if ([num1 integerValue] < [num2 integerValue]) {
                return NSOrderedAscending;
            } else if ([num1 integerValue] > [num2 integerValue]) {
                return NSOrderedDescending;
            } else {
                return NSOrderedSame;
            }
        };
        
        // Sort the array in ascending order using a comparator block
        [list1 sortUsingComparator:ascendingComparison];
        [list2 sortUsingComparator:ascendingComparison];
        
        NSInteger totalDistance = 0;
        for (int i = 0; i < (int)[list1 count]; i++) {
            totalDistance += labs([list1[i] integerValue] - [list2[i] integerValue]);
        }
        
        NSLog(@"%ld", totalDistance);
    }
    
    return 0;
}

File: day1/swift/day1-part1.swift
import Foundation

let directoryPath = FileManager.default.currentDirectoryPath

// Go back one directory (use ".." to navigate up)
let parentDirectoryPath = (directoryPath as NSString).deletingLastPathComponent  // Go one directory up
let filePath = parentDirectoryPath + "/input.txt"

let fileContents = try! String(contentsOfFile: filePath, encoding: .utf8)

let pairs = fileContents
    .components(separatedBy: "\n")
    .map { $0.split(separator: " ") }

var list1 = [Int]()
var list2 = [Int]()

for pair in pairs {
    list1.append(Int(pair[0])!)
    list2.append(Int(pair[1])!)
}

list1.sort()
list2.sort()

var totalDistance = 0
for i in 0..<list1.count {
    totalDistance += abs(list1[i] - list2[i])
}

print(totalDistance)
